syncer/sharding-meta/shardmeta.go

```diff
@@ -0,0 +1,237 @@
+	sources   map[string]*ShardingSequence // source table ID -> its sharding sequence
+}
+
+// IsPrefixSequence checks whether a ShardingSequence is the prefix sequence of other.
+func (seq *ShardingSequence) IsPrefixSequence(other *ShardingSequence) bool {
+	if len(seq.Items) > len(other.Items) {
+		return false
+	}
+	for idx := range seq.Items {
+		if !utils.CompareShardingDDLs(seq.Items[idx].DDLs, other.Items[idx].DDLs) {
+			return false
+		}
+	}
+	return true
+}
+
+// String returns the ShardingSequence's json string
+func (seq *ShardingSequence) String() string {
+	jsonSeq, err := json.Marshal(seq.Items)
+	if err != nil {
+		log.Errorf("marshal ShardingSequence to json error %v", err)
+	}
+	return string(jsonSeq)
+}
+
+// NewDDLItem creates a new DDLItem
+func NewDDLItem(pos mysql.Position, ddls []string, source string) *DDLItem {
+	return &DDLItem{
+		FirstPos: pos,
+		DDLs:     ddls,
+		Source:   source,
+	}
+}
+
+// NewShardingMeta creates a new ShardingMeta
+func NewShardingMeta() *ShardingMeta {
+	return &ShardingMeta{
+		global:  &ShardingSequence{Items: make([]*DDLItem, 0)},
+		sources: make(map[string]*ShardingSequence),
+	}
+}
+
+// LoadData loads data into ShardingMeta
+func (meta *ShardingMeta) LoadData(source string, activeIdx int, isGlobal bool, data []byte) error {
+	items := make([]*DDLItem, 0)
+	err := json.Unmarshal(data, &items)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if isGlobal {
+		meta.global = &ShardingSequence{Items: items}
+	} else {
+		meta.sources[source] = &ShardingSequence{Items: items}
+	}
+	meta.activeIdx = activeIdx
+	return nil
+}
+
+// ActiveIdx returns the activeIdx of sharding meta
+func (meta *ShardingMeta) ActiveIdx() int {
+	return meta.activeIdx
+}
+
+func (meta *ShardingMeta) reinitialize() {
+	meta.activeIdx = 0
+	meta.global = &ShardingSequence{make([]*DDLItem, 0)}
+	meta.sources = make(map[string]*ShardingSequence)
+}
+
+// checkItemExists checks whether DDLItem exists in its source sequence
+// if exists, return the index of DDLItem in source sequence.
+// if not exists, return the next index in source sequence.
+func (meta *ShardingMeta) checkItemExists(item *DDLItem) (int, bool) {
+	source, ok := meta.sources[item.Source]
+	if !ok {
+		return 0, false
+	}
+	for idx, ddlItem := range source.Items {
+		if item.FirstPos.Compare(ddlItem.FirstPos) == 0 {
+			return idx, true
+		}
+	}
+	return len(source.Items), false
+}
+
+// AddItem adds a new comming DDLItem into ShardingMeta
+// 1. if DDLItem already exists in source sequence, check whether it is active DDL only
+// 2. add the DDLItem into its related source sequence
+// 3. if it is a new DDL in global sequence, add it into global sequence
+// 4. check the source sequence is the prefix-sequence of global sequence, if not, return an error
+// returns:
+//   active: whether the DDL will be processed in this round
+func (meta *ShardingMeta) AddItem(item *DDLItem) (active bool, err error) {
```

Comment:
only have `AddItem` now, do we need add a function `RemoveItem` when ddl is synced?